### PR TITLE
[CLNP-5043] Migrate MessageSearchProvider

### PR DESCRIPTION
**Pull request by AhyoungRyu**
## Overview
This PR refactors the MessageSearch state management system, introducing a custom store solution and several hooks for improved performance, maintainability, and type safety.

## New Files
- `src/utils/storeManager.ts`: Implements a custom store creation utility
- `src/contexts/_MessageSearchContext.tsx`: Provides the MessageSearch context and store \w new state mgmt logic. It's just temporal name.
- `src/hooks/useStore.ts`: A generic hook for accessing and updating store state
- `src/hooks/useMessageSearchStore.ts`: A specialized hook for MessageSearch state management
- `src/components/MessageSearchManager.tsx`: Manages MessageSearch state and side effects
- `src/hooks/useMessageSearchActions.ts`: Manages action handlers 

## Updated Hooks
- `useSetChannel`: Now uses `useMessageSearchStore` directly
- `useSearchStringEffect`: Refactored to work with the new store
- `useGetSearchMessages`: Updated to utilize the new state management system
- `useScrollCallback`: Adapted to work with the custom store

## Key Changes
1. Introduced a custom store solution to replace the previous reducer-based state management.
2. Implemented `useStore` hook for type-safe and efficient state access and updates.
3. Created `MessageSearchManager` to centralize state management logic.
4. Refactored existing hooks to work with the new store system.
5. Improved type safety throughout the MessageSearch module.

### CODE REVIEW

File: src/hooks/useStore.ts [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+  // Ensure the stability of the selector function using useRef
+  const selectorRef = useRef(selector);
+  selectorRef.current = selector;
+  /**
+   * useSyncExternalStore - a new API introduced in React18
+   * useSyncExternalStore simply tracks changes in an external store that is not dependent on React
+   * through useState and useEffect
+   * and helps with re-rendering and state sync through the setter of useState
+   */
+  const state = useSyncExternalStore(
```
Comment (bang9) [inline src/hooks/useStore.ts:30]: It's due to our peerDependencies; we can't use `useSyncExternalStore` 🥲

```json
  "peerDependencies": {
    "react": "^16.8.6 || ^17.0.0 || ^18.0.0",
    "react-dom": "^16.8.6 || ^17.0.0 || ^18.0.0"
  },
```

Comment (bang9) [reply src/hooks/useStore.ts:30]: 👉 https://www.npmjs.com/package/use-sync-external-store

Comment (AhyoungRyu) [reply src/hooks/useStore.ts:32]: @bang9 Oh thank you for the catch! Just updated to use the 3rd party hook https://github.com/sendbird/sendbird-uikit-react/pull/1216/commits/f74c5713c50e17255eed0bc815b9d413196e2cae

**Comment by AhyoungRyu** [comment]
It seems there's an issue in Circle CI so the unit tests fail on the prepare stage. 
Screenshot 2024-09-27 at 1 41 41 PM

So I ran the test on my local machine. 
Screenshot 2024-09-27 at 1 41 15 PM

**Review by HoonBaek — APPROVED** [review]

**Pull request merged**